ICE:called `Result::unwrap()` on an `Err` value: NoSolution

_The code is generated by a fuzzer and reduced manually_
_I tried searching but couldn't find any related errors_
<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

### Code

```Rust
trait Trait<T> {}
struct X;
fn func<T>(x: *const dyn Trait<X>)
where
    for<'a> T_1: Sized,
{
    let x: *const dyn Trait<T> = x as _;
}
```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.89.0-nightly (70b3f4666 2025-05-30)
binary: rustc
commit-hash: 70b3f4666e24ce22fc32f5e357dbcf85d3254e63
commit-date: 2025-05-30
host: x86_64-pc-windows-msvc
release: 1.89.0-nightly
LLVM version: 20.1.5
```

### Error output

```
error[E0412]: cannot find type `T_1` in this scope
 --> .\test.rs:5:13
  |
5 |     for<'a> T_1: Sized,
  |             ^^^ not found in this scope
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
thread 'rustc' panicked at compiler\rustc_borrowck\src\type_check\mod.rs:1546:38:
called `Result::unwrap()` on an `Err` value: NoSolution
stack backtrace:
   0:     0x7ffe27e06082 - std::backtrace_rs::backtrace::win64::trace
                               at /rustc/70b3f4666e24ce22fc32f5e357dbcf85d3254e63/library\std\src\..\..\backtrace\src\backtrace\win64.rs:85
   1:     0x7ffe27e06082 - std::backtrace_rs::backtrace::trace_unsynchronized
                               at /rustc/70b3f4666e24ce22fc32f5e357dbcf85d3254e63/library\std\src\..\..\backtrace\src\backtrace\mod.rs:66
   2:     0x7ffe27e06082 - std::sys::backtrace::_print_fmt
                               at /rustc/70b3f4666e24ce22fc32f5e357dbcf85d3254e63/library\std\src\sys\backtrace.rs:66
   3:     0x7ffe27e06082 - std::sys::backtrace::impl$0::print::impl$0::fmt
                               at /rustc/70b3f4666e24ce22fc32f5e357dbcf85d3254e63/library\std\src\sys\backtrace.rs:39
   4:     0x7ffe27e39beb - core::fmt::rt::Argument::fmt
                               at /rustc/70b3f4666e24ce22fc32f5e357dbcf85d3254e63/library\core\src\fmt\rt.rs:173
   5:     0x7ffe27e39beb - core::fmt::write
                               at /rustc/70b3f4666e24ce22fc32f5e357dbcf85d3254e63/library\core\src\fmt\mod.rs:1446
   6:     0x7ffe27dfb9f7 - std::io::default_write_fmt
                               at /rustc/70b3f4666e24ce22fc32f5e357dbcf85d3254e63/library\std\src\io\mod.rs:639
   7:     0x7ffe27dfb9f7 - std::io::Write::write_fmt<std::sys::stdio::windows::Stderr>
                               at /rustc/70b3f4666e24ce22fc32f5e357dbcf85d3254e63/library\std\src\io\mod.rs:1914
   8:     0x7ffe27e05ec5 - std::sys::backtrace::BacktraceLock::print
                               at /rustc/70b3f4666e24ce22fc32f5e357dbcf85d3254e63/library\std\src\sys\backtrace.rs:42
   9:     0x7ffe27e0bc5a - std::panicking::default_hook::closure$0
                               at /rustc/70b3f4666e24ce22fc32f5e357dbcf85d3254e63/library\std\src\panicking.rs:300
  10:     0x7ffe27e0b9f0 - std::panicking::default_hook
                               at /rustc/70b3f4666e24ce22fc32f5e357dbcf85d3254e63/library\std\src\panicking.rs:327
  11:     0x7ffe295e4939 - core[41883027321c69d]::slice::sort::unstable::heapsort::heapsort::<((rustc_lint_defs[c6afabcd45b33dea]::Level, &str), usize), <((rustc_lint_defs[c6afabcd45b33dea]::Level, &str), usize) as core[41883027321c69d]::cmp::PartialOrd>::lt>
  12:     0x7ffe27e0c8de - std::panicking::rust_panic_with_hook
                               at /rustc/70b3f4666e24ce22fc32f5e357dbcf85d3254e63/library\std\src\panicking.rs:841
  13:     0x7ffe27e0c639 - std::panicking::begin_panic_handler::closure$0
                               at /rustc/70b3f4666e24ce22fc32f5e357dbcf85d3254e63/library\std\src\panicking.rs:706
  14:     0x7ffe27e06e4f - std::sys::backtrace::__rust_end_short_backtrace<std::panicking::begin_panic_handler::closure_env$0,never$>
                               at /rustc/70b3f4666e24ce22fc32f5e357dbcf85d3254e63/library\std\src\sys\backtrace.rs:168
  15:     0x7ffe27e0c23e - std::panicking::begin_panic_handler
                               at /rustc/70b3f4666e24ce22fc32f5e357dbcf85d3254e63/library\std\src\panicking.rs:697
  16:     0x7ffe2b032461 - core::panicking::panic_fmt
                               at /rustc/70b3f4666e24ce22fc32f5e357dbcf85d3254e63/library\core\src\panicking.rs:75
  17:     0x7ffe2b032a20 - core::result::unwrap_failed
                               at /rustc/70b3f4666e24ce22fc32f5e357dbcf85d3254e63/library\core\src\result.rs:1762
  18:     0x7ffe28679067 - <rustc_borrowck[e7ef4d8e0c81bd89]::type_check::TypeChecker as rustc_middle[7d96dcf9592acf23]::mir::visit::Visitor>::visit_rvalue
  19:     0x7ffe2862b4d8 - <rustc_borrowck[e7ef4d8e0c81bd89]::type_check::TypeChecker as rustc_middle[7d96dcf9592acf23]::mir::visit::Visitor>::visit_body
  20:     0x7ffe286b7380 - <rustc_borrowck[e7ef4d8e0c81bd89]::type_check::TypeChecker as rustc_middle[7d96dcf9592acf23]::mir::visit::Visitor>::visit_const_operand
  21:     0x7ffe285d927d - rustc_borrowck[e7ef4d8e0c81bd89]::mir_borrowck
  22:     0x7ffe28d6ae4b - rustc_query_impl[678ad264131b9188]::plumbing::query_key_hash_verify_all
  23:     0x7ffe28cb05a4 - RINvNtNtCs41YRoa67ONq_18rustc_query_system5query8plumbing17try_execute_queryINtCs8T9uthRx4cC_16rustc_query_impl13DynamicConfigINtNtCs2IPh3W6Af3f_21rustc_data_structures9vec_cache8VecCacheNtNtCsCuiHuraXL5_10rustc_span6def_id10LocalDefIdINtNtNtCsaMvyobqvdvz_
  24:     0x7ffe28d745d2 - rustc_query_impl[678ad264131b9188]::plumbing::query_key_hash_verify_all
  25:     0x7ffe27e7b084 - <rustc_interface[d66ed646d922d3e3]::passes::LintStoreExpandImpl as rustc_expand[32f7ad9713d55b7b]::base::LintStoreExpand>::pre_expansion_lint
  26:     0x7ffe24443d47 - rustc_interface[d66ed646d922d3e3]::passes::analysis
  27:     0x7ffe279ce36a - <alloc[40671bb55bd7e7eb]::sync::Arc<rustc_session[1833ef01fbdfa47f]::config::OutputFilenames>>::drop_slow
  28:     0x7ffe278d3c6a - RINvNtNtCs41YRoa67ONq_18rustc_query_system5query8plumbing17try_execute_queryINtCs8T9uthRx4cC_16rustc_query_impl13DynamicConfigINtNtB4_6caches11SingleCacheINtNtNtCsaMvyobqvdvz_12rustc_middle5query5erase6ErasedAhj0_EEKb0_KB3s_KB3s_ENtNtB1f_8plumbing9QueryCtx
  29:     0x7ffe279d4693 - rustc_query_impl[678ad264131b9188]::query_system
  30:     0x7ffe244091b4 - std[aa810135f6511839]::sys::backtrace::__rust_begin_short_backtrace::<<std[aa810135f6511839]::thread::Builder>::spawn_unchecked_<ctrlc[f2e6f1d10cfc7921]::set_handler_inner<rustc_driver_impl[e74b92f706932448]::install_ctrlc_handler::{closure#0}>::{closure#0}, ()>::{closure#1}::{closure#0}::{closure#0}, ()>
  31:     0x7ffe24403465 - std[aa810135f6511839]::sys::backtrace::__rust_begin_short_backtrace::<<std[aa810135f6511839]::thread::Builder>::spawn_unchecked_<ctrlc[f2e6f1d10cfc7921]::set_handler_inner<rustc_driver_impl[e74b92f706932448]::install_ctrlc_handler::{closure#0}>::{closure#0}, ()>::{closure#1}::{closure#0}::{closure#0}, ()>
  32:     0x7ffe243fed0f - RINvNtNtCseDAtLUMLJ79_3std3sys9backtrace28___rust_begin_short_backtraceNCNCINvNtCsipq8CQmOuQ1_15rustc_interface4util26run_in_thread_with_globalsNCINvB1e_31run_in_thread_pool_with_globalsNCINvNtB1g_9interface12run_compileruNCNvCsjRb60sNhVyu_17rustc_driver_i
  33:     0x7ffe2440d53d - std[aa810135f6511839]::sys::backtrace::__rust_begin_short_backtrace::<<std[aa810135f6511839]::thread::Builder>::spawn_unchecked_<ctrlc[f2e6f1d10cfc7921]::set_handler_inner<rustc_driver_impl[e74b92f706932448]::install_ctrlc_handler::{closure#0}>::{closure#0}, ()>::{closure#1}::{closure#0}::{closure#0}, ()>
  34:     0x7ffe27e10c3d - alloc::boxed::impl$28::call_once
                               at /rustc/70b3f4666e24ce22fc32f5e357dbcf85d3254e63/library\alloc\src\boxed.rs:1966
  35:     0x7ffe27e10c3d - alloc::boxed::impl$28::call_once
                               at /rustc/70b3f4666e24ce22fc32f5e357dbcf85d3254e63/library\alloc\src\boxed.rs:1966
  36:     0x7ffe27e10c3d - std::sys::pal::windows::thread::impl$0::new::thread_start
                               at /rustc/70b3f4666e24ce22fc32f5e357dbcf85d3254e63/library\std\src\sys\pal\windows\thread.rs:56
  37:     0x7ffedfd37374 - BaseThreadInitThunk
  38:     0x7ffee13bcc91 - RtlUserThreadStart

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

query stack during panic:
#0 [mir_borrowck] borrow-checking `func`
#1 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 1 previous error; 1 warning emitted
```

</p>
</details>


<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_borrowck/src/type_check/mod.rs
+++ b/compiler/rustc_borrowck/src/type_check/mod.rs
@@ -34,6 +34,7 @@ use rustc_mir_dataflow::points::DenseLocationMap;
 use rustc_span::def_id::CRATE_DEF_ID;
 use rustc_span::source_map::Spanned;
 use rustc_span::{Span, sym};
+use rustc_trait_selection::infer::InferCtxtExt;
 use rustc_trait_selection::traits::query::type_op::custom::scrape_region_constraints;
 use rustc_trait_selection::traits::query::type_op::{TypeOp, TypeOpOutput};
 use tracing::{debug, instrument, trace};
@@ -1457,22 +1458,32 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
                         let ty_from = op.ty(self.body, tcx);
                         let cast_ty_from = CastTy::from_ty(ty_from);
                         let cast_ty_to = CastTy::from_ty(*ty);
+
                         match (cast_ty_from, cast_ty_to) {
                             (Some(CastTy::Ptr(src)), Some(CastTy::Ptr(dst))) => {
-                                let src_tail = self.struct_tail(src.ty, location);
-                                let dst_tail = self.struct_tail(dst.ty, location);
-
-                                // This checks (lifetime part of) vtable validity for pointer casts,
-                                // which is irrelevant when there are aren't principal traits on
-                                // both sides (aka only auto traits).
-                                //
-                                // Note that other checks (such as denying `dyn Send` -> `dyn
-                                // Debug`) are in `rustc_hir_typeck`.
-                                if let ty::Dynamic(src_tty, _src_lt, ty::Dyn) = *src_tail.kind()
-                                    && let ty::Dynamic(dst_tty, dst_lt, ty::Dyn) = *dst_tail.kind()
+                                if self
+                                    .infcx
+                                    .type_is_sized_modulo_regions(self.infcx.param_env, dst.ty)
+                                {
+                                    // Wide to thin ptr cast. This may even occur in an env with
+                                    // impossible predicates, such as `where dyn Trait: Sized`.
+                                    // In this case, we don't want to fall into the case below,
+                                    // since the types may not actually be equatable, but it's
+                                    // fine to perform this operation in an impossible env.
+                                } else if let ty::Dynamic(src_tty, _src_lt, ty::Dyn) =
+                                    *self.struct_tail(src.ty, location).kind()
+                                    && let ty::Dynamic(dst_tty, dst_lt, ty::Dyn) =
+                                        *self.struct_tail(dst.ty, location).kind()
                                     && src_tty.principal().is_some()
                                     && dst_tty.principal().is_some()
                                 {
+                                    // This checks (lifetime part of) vtable validity for pointer casts,
+                                    // which is irrelevant when there are aren't principal traits on
+                                    // both sides (aka only auto traits).
+                                    //
+                                    // Note that other checks (such as denying `dyn Send` -> `dyn
+                                    // Debug`) are in `rustc_hir_typeck`.
+
                                     // Remove auto traits.
                                     // Auto trait checks are handled in `rustc_hir_typeck` as FCW.
                                     let src_obj = Ty::new_dynamic(

--- a/compiler/rustc_borrowck/src/type_check/mod.rs
+++ b/compiler/rustc_borrowck/src/type_check/mod.rs
@@ -1456,78 +1456,79 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
                     }
                     CastKind::PtrToPtr => {
                         let ty_from = op.ty(self.body, tcx);
-                        let cast_ty_from = CastTy::from_ty(ty_from);
-                        let cast_ty_to = CastTy::from_ty(*ty);
+                        let Some(CastTy::Ptr(src)) = CastTy::from_ty(ty_from) else {
+                            unreachable!();
+                        };
+                        let Some(CastTy::Ptr(dst)) = CastTy::from_ty(*ty) else {
+                            unreachable!();
+                        };
 
-                        match (cast_ty_from, cast_ty_to) {
-                            (Some(CastTy::Ptr(src)), Some(CastTy::Ptr(dst))) => {
-                                if self
-                                    .infcx
-                                    .type_is_sized_modulo_regions(self.infcx.param_env, dst.ty)
-                                {
-                                    // Wide to thin ptr cast. This may even occur in an env with
-                                    // impossible predicates, such as `where dyn Trait: Sized`.
-                                    // In this case, we don't want to fall into the case below,
-                                    // since the types may not actually be equatable, but it's
-                                    // fine to perform this operation in an impossible env.
-                                } else if let ty::Dynamic(src_tty, _src_lt, ty::Dyn) =
-                                    *self.struct_tail(src.ty, location).kind()
-                                    && let ty::Dynamic(dst_tty, dst_lt, ty::Dyn) =
-                                        *self.struct_tail(dst.ty, location).kind()
-                                    && src_tty.principal().is_some()
-                                    && dst_tty.principal().is_some()
-                                {
-                                    // This checks (lifetime part of) vtable validity for pointer casts,
-                                    // which is irrelevant when there are aren't principal traits on
-                                    // both sides (aka only auto traits).
-                                    //
-                                    // Note that other checks (such as denying `dyn Send` -> `dyn
-                                    // Debug`) are in `rustc_hir_typeck`.
+                        if self.infcx.type_is_sized_modulo_regions(self.infcx.param_env, dst.ty) {
+                            // Wide to thin ptr cast. This may even occur in an env with
+                            // impossible predicates, such as `where dyn Trait: Sized`.
+                            // In this case, we don't want to fall into the case below,
+                            // since the types may not actually be equatable, but it's
+                            // fine to perform this operation in an impossible env.
+                            let trait_ref = ty::TraitRef::new(
+                                tcx,
+                                tcx.require_lang_item(LangItem::Sized, self.last_span),
+                                [dst.ty],
+                            );
+                            self.prove_trait_ref(
+                                trait_ref,
+                                location.to_locations(),
+                                ConstraintCategory::Cast {
+                                    is_implicit_coercion: true,
+                                    unsize_to: None,
+                                },
+                            );
+                        } else if let ty::Dynamic(src_tty, _src_lt, ty::Dyn) =
+                            *self.struct_tail(src.ty, location).kind()
+                            && let ty::Dynamic(dst_tty, dst_lt, ty::Dyn) =
+                                *self.struct_tail(dst.ty, location).kind()
+                            && src_tty.principal().is_some()
+                            && dst_tty.principal().is_some()
+                        {
+                            // This checks (lifetime part of) vtable validity for pointer casts,
+                            // which is irrelevant when there are aren't principal traits on
+                            // both sides (aka only auto traits).
+                            //
+                            // Note that other checks (such as denying `dyn Send` -> `dyn
+                            // Debug`) are in `rustc_hir_typeck`.
 
-                                    // Remove auto traits.
-                                    // Auto trait checks are handled in `rustc_hir_typeck` as FCW.
-                                    let src_obj = Ty::new_dynamic(
-                                        tcx,
-                                        tcx.mk_poly_existential_predicates(
-                                            &src_tty.without_auto_traits().collect::<Vec<_>>(),
-                                        ),
-                                        // FIXME: Once we disallow casting `*const dyn Trait + 'short`
-                                        // to `*const dyn Trait + 'long`, then this can just be `src_lt`.
-                                        dst_lt,
-                                        ty::Dyn,
-                                    );
-                                    let dst_obj = Ty::new_dynamic(
-                                        tcx,
-                                        tcx.mk_poly_existential_predicates(
-                                            &dst_tty.without_auto_traits().collect::<Vec<_>>(),
-                                        ),
-                                        dst_lt,
-                                        ty::Dyn,
-                                    );
+                            // Remove auto traits.
+                            // Auto trait checks are handled in `rustc_hir_typeck` as FCW.
+                            let src_obj = Ty::new_dynamic(
+                                tcx,
+                                tcx.mk_poly_existential_predicates(
+                                    &src_tty.without_auto_traits().collect::<Vec<_>>(),
+                                ),
+                                // FIXME: Once we disallow casting `*const dyn Trait + 'short`
+                                // to `*const dyn Trait + 'long`, then this can just be `src_lt`.
+                                dst_lt,
+                                ty::Dyn,
+                            );
+                            let dst_obj = Ty::new_dynamic(
+                                tcx,
+                                tcx.mk_poly_existential_predicates(
+                                    &dst_tty.without_auto_traits().collect::<Vec<_>>(),
+                                ),
+                                dst_lt,
+                                ty::Dyn,
+                            );
 
-                                    debug!(?src_tty, ?dst_tty, ?src_obj, ?dst_obj);
+                            debug!(?src_tty, ?dst_tty, ?src_obj, ?dst_obj);
 
-                                    self.sub_types(
-                                        src_obj,
-                                        dst_obj,
-                                        location.to_locations(),
-                                        ConstraintCategory::Cast {
-                                            is_implicit_coercion: false,
-                                            unsize_to: None,
-                                        },
-                                    )
-                                    .unwrap();
-                                }
-                            }
-                            _ => {
-                                span_mirbug!(
-                                    self,
-                                    rvalue,
-                                    "Invalid PtrToPtr cast {:?} -> {:?}",
-                                    ty_from,
-                                    ty
-                                )
-                            }
+                            self.sub_types(
+                                src_obj,
+                                dst_obj,
+                                location.to_locations(),
+                                ConstraintCategory::Cast {
+                                    is_implicit_coercion: false,
+                                    unsize_to: None,
+                                },
+                            )
+                            .unwrap();
                         }
                     }
                     CastKind::Transmute => {

--- a/tests/ui/coercion/fake-sized-ptr-cast.rs
+++ b/tests/ui/coercion/fake-sized-ptr-cast.rs
@@ -1,0 +1,16 @@
+// Make sure borrowck doesn't ICE because it thinks a pointer cast is a metadata-preserving
+// wide-to-wide ptr cast when it's actually (falsely) a wide-to-thin ptr cast due to an
+// impossible dyn sized bound.
+
+//@ check-pass
+
+trait Trait<T> {}
+
+fn func<'a>(x: *const (dyn Trait<()> + 'a))
+where
+    dyn Trait<u8> + 'a: Sized,
+{
+    let _x: *const dyn Trait<u8> = x as _;
+}
+
+fn main() {}

